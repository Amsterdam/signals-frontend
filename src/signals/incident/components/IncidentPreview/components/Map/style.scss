.preview-map {
  &__item-label,
  &__item-value {
    padding: 20px 0;

    &-map {
      margin-top: 8px;

      #mapdiv {
        height: 300px;
      }
    }
  }
}
