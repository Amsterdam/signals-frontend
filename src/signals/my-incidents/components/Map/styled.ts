// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2022 Gemeente Amsterdam
import { themeSpacing } from '@amsterdam/asc-ui'
import styled from 'styled-components'

import CloseButton from 'components/CloseButton'

import MapDetail from '../../../../components/MapDetail'

export const StyledMapViewer = styled.div`
  position: relative;
  top: ${themeSpacing(5)};
  width: 100%;
`

export const StyledCloseButton = styled(CloseButton)`
  z-index: 400;
  top: 16px;
  right: 0;
`

export const StyledMapDetail = styled(MapDetail)`
  height: 450px;
  width: 100%;
`
