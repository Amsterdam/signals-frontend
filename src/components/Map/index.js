import React from 'react';
import PropTypes from 'prop-types';

import amaps from '../../static/amaps.iife';

import './style.scss';

const PREVIEW_ZOOM_LEVEL = 16;

class Map extends React.Component {
  componentDidMount() {
    amaps.createMap({
      center: {
        latitude: this.props.latlng.latitude,
        longitude: this.props.latlng.longitude
      },
      layer: 'standaard',
      target: 'mapdiv',
      marker: true,
      search: false,
      zoom: PREVIEW_ZOOM_LEVEL
    });
  }

  render() {
    return (
      <div className="map-component">
        <div className="map">
          <div id="mapdiv" />
        </div>
      </div>
    );
  }
}

Map.defaultProps = {
  latlng: {}
};

Map.propTypes = {
  latlng: PropTypes.object
};

export default Map;
