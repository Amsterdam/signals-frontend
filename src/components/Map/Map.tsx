// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2019 - 2021 Gemeente Amsterdam
import type { FC, PropsWithChildren } from 'react'
import { useEffect } from 'react'

import { Zoom, Map as MapComponent } from '@amsterdam/arm-core'
import { TileLayer } from '@amsterdam/react-maps'
import type { LeafletEventHandlerFnMap, MapOptions } from 'leaflet'
import { useDispatch, useSelector } from 'react-redux'
import styled from 'styled-components'

import ViewerContainer from 'components/ViewerContainer'
import configuration from 'shared/services/configuration/configuration'
import { closeMap } from 'signals/incident/containers/IncidentContainer/actions'
import { makeSelectIncidentContainer } from 'signals/incident/containers/IncidentContainer/selectors'

const StyledMap = styled(MapComponent)`
  cursor: default;

  &:focus {
    outline: 5px auto Highlight !important; // Firefox outline
    outline: 5px auto -webkit-focus-ring-color !important; // Safari / Chrome outline
  }

  &.leaflet-drag-target {
    cursor: all-scroll;
  }
`

export interface MapProps {
  className?: string
  'data-testid'?: string
  /**
   * Map events
   * @see {@link https://leafletjs.com/reference-1.6.0.html#map-event}
   */
  events?: LeafletEventHandlerFnMap
  hasGPSControl?: boolean
  hasZoomControls?: boolean
  /**
   * Leaflet configuration options
   * @see {@link https://leafletjs.com/reference-1.6.0.html#map-option}
   */
  mapOptions: MapOptions
  fullScreen?: boolean
  /**
   * useState function that sets a reference to the map instance
   */
  setInstance?: (instance: L.Map) => void
}

const Map: FC<PropsWithChildren<MapProps>> = ({
  'data-testid': dataTestId = 'map-base',
  children,
  className = '',
  events,
  fullScreen = false,
  hasZoomControls = false,
  mapOptions,
  setInstance,
}) => {
  const hasTouchCapabilities = 'ontouchstart' in window
  const showZoom = hasZoomControls && !hasTouchCapabilities
  const maxZoom = mapOptions?.maxZoom || configuration.map.options.maxZoom
  const minZoom = mapOptions?.minZoom || configuration.map.options.minZoom
  const options = {
    maxZoom,
    minZoom,
    tap: false,
    scrollWheelZoom: false,
    center: mapOptions?.center,
    keyboard: false,
    ...mapOptions,
  }

  const dispatch = useDispatch()
  const { mapActive } = useSelector(makeSelectIncidentContainer)

  useEffect(() => {
    return () => {
      if (mapActive) {
        dispatch(closeMap())
      }
    }
  }, [mapActive, dispatch])

  return (
    <StyledMap
      // Disabling linter; without className prop, the Map component cannot be styled
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      className={className}
      data-testid={dataTestId}
      data-max-zoom={maxZoom}
      data-min-zoom={minZoom}
      events={events}
      options={options}
      fullScreen={fullScreen}
      setInstance={setInstance}
    >
      {children}

      {/* Render zoom buttons after children to maintain correct focus order */}
      <ViewerContainer
        bottomRight={
          showZoom && (
            <div data-testid="map-zoom">
              <Zoom />
            </div>
          )
        }
      />

      <TileLayer
        args={configuration.map.tiles.args as [string]}
        options={configuration.map.tiles.options}
      />
    </StyledMap>
  )
}

export default Map
