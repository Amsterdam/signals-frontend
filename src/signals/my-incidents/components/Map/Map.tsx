// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2022 Gemeente Amsterdam
import configuration from 'shared/services/configuration/configuration'
import type Location from 'types/location'

import { StyledCloseButton, StyledMapDetail, StyledMapViewer } from './styled'
import { markerIcon } from '../../../../shared/services/configuration/map-markers'

type Props = {
  location: Location
  close: () => void
}

export const Map = ({ location, close }: Props) => (
  <StyledMapViewer>
    <StyledMapDetail
      value={location}
      icon={markerIcon}
      hasZoomControls
      zoom={configuration.map.optionsMyIncidentsMap.zoom}
    />
    <StyledCloseButton close={close} />
  </StyledMapViewer>
)
