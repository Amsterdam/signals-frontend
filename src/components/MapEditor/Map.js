import React from 'react';
import PropTypes from 'prop-types';
import { ViewerContainer } from '@datapunt/asc-ui';
import { Zoom } from '@datapunt/amsterdam-react-maps/lib/components';
import styled from '@datapunt/asc-core';
import { Map as MapComponent, TileLayer } from '@datapunt/react-maps';

const Wrapper = styled(MapComponent)`
  height: 450px;
  width: 100%;
`;

const StyledViewerContainer = styled(ViewerContainer)`
  z-index: 400; // this elevation ensures that this container comes on top of the internal leaflet components
`;

const hasTouchCapabilities = !!global.L.Browser.touch;

const Map = ({ mapOptions, hasZoomControls, hasAttributionControl, isInteractive, children, ...otherProps }) => {
  const options = {
    ...mapOptions,
    attributionControl: hasAttributionControl,
    dragging: isInteractive && !hasTouchCapabilities,
    tap: isInteractive && !hasTouchCapabilities,
    scrollWheelZoom: isInteractive && !hasTouchCapabilities,
  };

  return (
    <Wrapper data-testid="map-test-id" options={options} {...otherProps}>
      {hasZoomControls && <StyledViewerContainer bottomRight={<Zoom />} />}

      {children}

      <TileLayer
        args={['https://{s}.data.amsterdam.nl/topo_rd/{z}/{x}/{y}.png']}
        options={{
          subdomains: ['t1', 't2', 't3', 't4'],
          tms: true,
          attribution: hasAttributionControl && 'Kaartgegevens CC-BY-4.0 Gemeente Amsterdam',
        }}
      />
    </Wrapper>
  );
};
Map.defaultProps = {
  hasZoomControls: false,
  isInteractive: true,
  hasAttributionControl: true,
};

Map.propTypes = {
  mapOptions: PropTypes.shape({
    attributionControl: PropTypes.bool,
  }).isRequired /** leaflet options, See `https://leafletjs.com/reference-1.6.0.html#map-option` */,
  hasZoomControls: PropTypes.bool,
  hasAttributionControl: PropTypes.bool,
  isInteractive: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
};

export default Map;
