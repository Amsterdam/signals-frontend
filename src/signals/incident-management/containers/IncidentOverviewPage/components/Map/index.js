import React, { memo, useContext, useState, useCallback, useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { useSelector } from 'react-redux';
import isEqual from 'lodash.isequal';
import format from 'date-fns/format';
import subDays from 'date-fns/addDays';
import L from 'leaflet';
import { ViewerContainer, themeColor, themeSpacing } from '@datapunt/asc-ui';

import MapContext from 'containers/MapContext/context';
import { setAddressAction } from 'containers/MapContext/actions';
import MAP_OPTIONS from 'shared/services/configuration/map-options';
import configuration from 'shared/services/configuration/configuration';
import { featureTolocation, formatPDOKResponse } from 'shared/services/map-location';
import { makeSelectFilterParams, makeSelectActiveFilter } from 'signals/incident-management/selectors';
import { initialState } from 'signals/incident-management/reducer';
import useFetch from 'hooks/useFetch';
import { incidentIcon, markerIcon } from 'shared/services/configuration/map-markers';
import Map from 'components/Map';
import PDOKAutoSuggest from 'components/PDOKAutoSuggest';
import MarkerCluster from './components/MarkerCluster';

import DetailPanel from './components/DetailPanel';

const StyledViewerContainer = styled(ViewerContainer)`
  flex-direction: row;

  & > * {
    left: ${themeSpacing(4)};
    right: ${themeSpacing(4)};
  }
`;

const Wrapper = styled.div`
  position: relative;
  width: 100%;
`;

const StyledMap = styled(Map)`
  height: 600px;
  width: 100%;

  .marker-cluster {
    color: ${themeColor('tint', 'level1')};
    background-color: ${themeColor('tint', 'level1')};
    box-shadow: 1px 1px 1px #666666;

    div {
      width: 32px;
      height: 32px;
      margin-top: 4px;
      margin-left: 4px;
      background-color: #004699;
    }

    span {
      line-height: 34px;
    }
  }
`;

const Autosuggest = styled(PDOKAutoSuggest)`
  max-width: calc(100% - 40px);
  z-index: 401; // 400 is the minimum elevation were elements are shown above the map
  width: 350px;
  left: 0;
  position: absolute;
`;

const clusterLayerOptions = {
  showCoverageOnHover: false,
  zoomToBoundsOnClick: true,
  chunkedLoading: true,
  iconCreateFunction: cluster => {
    const childCount = cluster.getChildCount();
    let c = ' marker-cluster-';

    if (childCount < 10) {
      c += 'small';
    } else if (childCount < 100) {
      c += 'medium';
    } else {
      c += 'large';
    }

    return new L.DivIcon({
      html: `<div data-testid="markerClusterIcon"><span>${childCount}</span></div>`,
      className: `marker-cluster ${c}`,
      iconSize: new L.Point(40, 40),
    });
  },
};

const OverviewMap = ({ showPanelOnInit, ...rest }) => {
  const { dispatch } = useContext(MapContext);
  const [initialMount, setInitialMount] = useState(false);
  const [showPanel, setShowPanel] = useState(showPanelOnInit);
  const [map, setMap] = useState();
  const { options } = useSelector(makeSelectActiveFilter);
  const filterParams = useSelector(makeSelectFilterParams);
  const { get, data, isLoading } = useFetch();
  const [layerInstance, setLayerInstance] = useState();
  const [incidentId, setIncidentId] = useState(0);

  const { ...params } = filterParams;

  // fixed query period (24 hours)
  params.created_after = useMemo(() => format(subDays(new Date(), -1), "yyyy-MM-dd'T'HH:mm:ss"), []);
  params.created_before = useMemo(() => format(new Date(), "yyyy-MM-dd'T'HH:mm:ss"), []);

  // fixed page size (default is 50; 4000 is 2.5 times the highest daily average)
  params.page_size = 4000;

  /**
   * AutoSuggest callback handler
   *
   * Note that testing this functionality resembles integration testing, hence disabling istanbul coverage
   */
  const onSelect = useCallback(
    /* istanbul ignore next */ option => {
      dispatch(setAddressAction(option.value));

      if (map) {
        const currentZoom = map.getZoom();
        map.flyTo(option.data.location, currentZoom < 11 ? 11 : currentZoom);
      }
    },
    [map, dispatch]
  );

  const resetMarkerIcons = useCallback(() => {
    map.eachLayer(layer => {
      if (layer.getIcon && !layer.getAllChildMarkers) {
        layer.setIcon(incidentIcon);
      }
    });
  }, [map]);

  const onClosePanel = useCallback(() => {
    setShowPanel(false);
    resetMarkerIcons();
  }, [resetMarkerIcons]);

  useEffect(() => {
    if (!options || isLoading || !initialMount) return;

    const { name, ...initialActive } = initialState.get('activeFilter').toJS();
    const paramsAreInitial = isEqual(initialActive.options, options);

    if (paramsAreInitial) return;

    get(`${configuration.GEOGRAPHY_ENDPOINT}`, params);

    // Only execute when the value of filterParams changes; disabling linter
    // eslint-disable-next-line
  }, [filterParams]);

  // request data on mount
  useEffect(() => {
    get(`${configuration.GEOGRAPHY_ENDPOINT}`, params);
    setInitialMount(true);

    // eslint-disable-next-line
  }, []);

  useEffect(() => {
    if (!data || !layerInstance) return () => {};

    data.features.forEach(feature => {
      const latlng = featureTolocation(feature.geometry);

      const clusteredMarker = L.marker(latlng, {
        icon: incidentIcon,
      });

      clusteredMarker.on('click', event => {
        resetMarkerIcons();

        event.target.setIcon(markerIcon);

        /* istanbul ignore else */
        if (feature.properties?.id) {
          setIncidentId(feature.properties.id);
          setShowPanel(true);
        }
      });

      layerInstance.addLayer(clusteredMarker);
    });

    return () => {
      layerInstance.clearLayers();
    };
  }, [layerInstance, data, map, resetMarkerIcons]);

  return (
    <Wrapper {...rest}>
      <StyledMap
        data-testid="overviewMap"
        hasZoomControls
        mapOptions={{
          ...MAP_OPTIONS,
          ...(configuration.map.optionsBackOffice || {}),
        }}
        setInstance={setMap}
      >
        <MarkerCluster clusterOptions={clusterLayerOptions} setInstance={setLayerInstance} />
        <StyledViewerContainer
          topLeft={
            <Autosuggest
              fieldList={['centroide_ll']}
              formatResponse={formatPDOKResponse}
              municipality={configuration.map?.municipality}
              onSelect={onSelect}
              placeholder="Zoom naar adres"
            />
          }
          topRight={showPanel && <DetailPanel incidentId={incidentId} onClose={onClosePanel} />}
        />
      </StyledMap>
    </Wrapper>
  );
};

OverviewMap.defaultProps = {
  showPanelOnInit: false,
};

OverviewMap.propTypes = {
  showPanelOnInit: PropTypes.bool,
};

export default memo(OverviewMap);
