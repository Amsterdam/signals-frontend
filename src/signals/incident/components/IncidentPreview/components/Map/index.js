import React from 'react';
import PropTypes from 'prop-types';
import Map from '../../../../../../components/Map';

import './style.scss';

const MapPreview = ({ label, value, optional }) => (
  <div className="preview-map">
    {!optional || (optional && value) ?
      <div className="row">
        <div className="col-5 col-md-4">
          <div className="preview-map__item-label">{label}</div>
        </div>
        <div className="col-5 col-md-7">
          <div className="preview-map__item-value">
            {value ?
              <div>
                <div>
                  {value.address ? `${value.address.openbare_ruimte} ${value.address.huisnummer}${value.address.huisletter}` : 'geen adres gevonden'}</div>
                {value.geometrie && value.geometrie.coordinates ?
                  <div className="preview-map__item-value-map">
                    <Map
                      latlng={{ latitude: value.geometrie.coordinates[1], longitude: value.geometrie.coordinates[0] }}
                    />
                  </div>
                : ''}
              </div>
              : ''}
          </div>
        </div>
      </div>
      : ''}
  </div>
);

MapPreview.propTypes = {
  label: PropTypes.string,
  value: PropTypes.object,
  optional: PropTypes.bool
};

export default MapPreview;
