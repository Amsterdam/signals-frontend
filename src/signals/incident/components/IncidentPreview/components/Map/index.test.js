import React from 'react';
import { shallow } from 'enzyme';

import Map from './index';

describe('Preview component <Map />', () => {
  const address = {
    openbare_ruimte: 'Hell',
    huisnummer: '666',
    huisletter: 'C'
  };
  const geometrie = {
    coordinates: [52, 4]
  };
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Map />);
  });

  it('should render valid adrress and map with latlong correctly', () => {
    wrapper.setProps({
      label: 'Location',
      value: {
        address,
        geometrie
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render only latlong correctly', () => {
    wrapper.setProps({
      label: 'Location',
      value: {
        geometrie
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render only address correctly', () => {
    wrapper.setProps({
      label: 'Location',
      value: {
        address
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render optional map correctly', () => {
    wrapper.setProps({
      label: 'Location',
      value: {
        address,
        geometrie
      },
      optional: true
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should not render optional map when value is empty', () => {
    wrapper.setProps({
      label: 'Location',
      optional: true
    });

    expect(wrapper).toMatchSnapshot();
  });
});
