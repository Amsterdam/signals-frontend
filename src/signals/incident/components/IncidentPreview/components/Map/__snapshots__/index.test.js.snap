// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview component <Map /> should not render optional map when value is empty 1`] = `
<div
  className="preview-map"
/>
`;

exports[`Preview component <Map /> should render only address correctly 1`] = `
<div
  className="preview-map"
>
  <div
    className="row"
  >
    <div
      className="col-5 col-md-4"
    >
      <div
        className="preview-map__item-label"
      >
        Location
      </div>
    </div>
    <div
      className="col-5 col-md-7"
    >
      <div
        className="preview-map__item-value"
      >
        <div>
          <div>
            Hell 666C
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Preview component <Map /> should render only latlong correctly 1`] = `
<div
  className="preview-map"
>
  <div
    className="row"
  >
    <div
      className="col-5 col-md-4"
    >
      <div
        className="preview-map__item-label"
      >
        Location
      </div>
    </div>
    <div
      className="col-5 col-md-7"
    >
      <div
        className="preview-map__item-value"
      >
        <div>
          <div>
            geen adres gevonden
          </div>
          <div
            className="preview-map__item-value-map"
          >
            <Map
              latlng={
                Object {
                  "latitude": 4,
                  "longitude": 52,
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Preview component <Map /> should render optional map correctly 1`] = `
<div
  className="preview-map"
>
  <div
    className="row"
  >
    <div
      className="col-5 col-md-4"
    >
      <div
        className="preview-map__item-label"
      >
        Location
      </div>
    </div>
    <div
      className="col-5 col-md-7"
    >
      <div
        className="preview-map__item-value"
      >
        <div>
          <div>
            Hell 666C
          </div>
          <div
            className="preview-map__item-value-map"
          >
            <Map
              latlng={
                Object {
                  "latitude": 4,
                  "longitude": 52,
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Preview component <Map /> should render valid adrress and map with latlong correctly 1`] = `
<div
  className="preview-map"
>
  <div
    className="row"
  >
    <div
      className="col-5 col-md-4"
    >
      <div
        className="preview-map__item-label"
      >
        Location
      </div>
    </div>
    <div
      className="col-5 col-md-7"
    >
      <div
        className="preview-map__item-value"
      >
        <div>
          <div>
            Hell 666C
          </div>
          <div
            className="preview-map__item-value-map"
          >
            <Map
              latlng={
                Object {
                  "latitude": 4,
                  "longitude": 52,
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
